dms/models/mixins_thumbnail.py

```diff
@@ -14,79 +14,86 @@ class Thumbnail(models.AbstractModel):
-    def _get_thumbnail_placeholder_image(self, size, name):
-        path = self._get_thumbnail_path(size, name)
+    def _get_thumbnail_placeholder_image(self, name):
+        path = self._get_thumbnail_path(name)
         with open(path, "rb") as image:
             return base64.b64encode(image.read())
 
     @api.model
-    def _get_thumbnail_path(self, size, name):
+    def _get_thumbnail_path(self, name):
         folders = ["static", "lib", "img", "thumbnails"]
         path = get_resource_path("dms", *folders, name)
         if not os.path.isfile(path):
-            path = get_resource_path("dms", *folders, "file_unkown.svg")
+            path = get_resource_path("dms", *folders, "file_unknown.svg")
         return path
 
     def _get_thumbnail_placeholder_name(self):
         return "folder.svg"
 
-    @api.depends("custom_thumbnail")
+    def _compute_icon_url(self):
+        """Get icon static file URL."""
+        for one in self:
+            thumbnail_path = one._get_thumbnail_path(
+                one._get_thumbnail_placeholder_name()
+            )
+            thumbnail_name = os.path.basename(thumbnail_path)
+            one.icon_url = "/dms/static/lib/img/thumbnails/%s" % thumbnail_name
+
+    @api.depends("custom_thumbnail_medium", "custom_thumbnail_small")
     def _compute_thumbnail(self):
+        """Generate small thumbnails, with icon fallback.
+
+        NEVER use ``thumbnail_*`` fields in list/kanban views, or things will
+        slow down. Instead, use directly ``icon_url`` or ``custom_thumbnail_*``.
+
+        It reflects ``custom_thumbnail_*`` fields if possible, but defaults to
+        the mimetype icon.
+        """
+        for record in self:
+            icon_fallback = record._get_thumbnail_placeholder_image(
+                record._get_thumbnail_placeholder_name()
+            )
+            for field in ("thumbnail", "thumbnail_medium", "thumbnail_small"):
+                record[field] = record["custom_%s" % field] or icon_fallback
+
+    @api.depends("custom_thumbnail")
+    def _compute_custom_thumbnail(self):
+        """Generate small custom thumbnails.
+
+        They will be empty if the main custom thumbnail is empty.
+        """
         for record in self:
-            if record.custom_thumbnail:
-                record.thumbnail = record.custom_thumbnail
-            else:
-                record.thumbnail = self._get_thumbnail_placeholder_image(
-                    "original", record._get_thumbnail_placeholder_name()
-                )
+            base64_source = record.with_context(bin_size=False).custom_thumbnail
+            record.update(
+                {
+                    "custom_thumbnail_medium": tools.image_process(
+                        base64_source, size=(128, 128)
+                    ),
+                    "custom_thumbnail_small": tools.image_process(
+                        base64_source, size=(64, 64)
+                    ),
+                }
+            )
```

Comment:
Probably it would be better to remove this function and use the max_width and max_height field attributes as they were being used (just modifying values). That's a new feature that didn't exist in v12 and thus I wasn't able to do it that way.

Please push that separately so it can be easily analyzed apart from this cherry-pick.